### PR TITLE
(PUP-11127) Readable type errors

### PR DESCRIPTION
**Pull request by artlawson**
These changes provide easier to read error messages for type errors 

Addressing (Bolt ticket ([2853](https://github.com/puppetlabs/bolt/issues/2853))

### CODE REVIEW

File: lib/puppet/pops/types/type_mismatch_describer.rb [hunk 1]
```diff
@@ -363,8 +363,11 @@ def specialization(e, a)
     def always_fully_detailed?(e, a)
       if e.is_a?(Array)
         e.any? { |t| always_fully_detailed?(t, a) }
+
+      elsif e.is_a?(Puppet::Pops::Types::PObjectType) && a.is_a?(Puppet::Pops::Types::PObjectType)
+        e.implementation_class(false) == a.implementation_class(false)
       else
-        e.class == a.class || e.is_a?(PTypeAliasType) || a.is_a?(PTypeAliasType) || specialization(e, a)
+        e.class == a.class ||  a.is_a?(PTypeAliasType) || specialization(e, a)
```
Comment (joshcooper) [inline lib/puppet/pops/types/type_mismatch_describer.rb:370]: I think we want to keep the check for `e.is_a?(PTypeAliasType)`

File: spec/unit/pops/types/type_mismatch_describer_spec.rb [hunk 2]
```diff
@@ -181,6 +184,59 @@ module Types
...
+        }
+      end
+    
+      before(:each) do
+        Puppet[:environment] = 'testing'
+      end
+      it "prints the short message when expected value is a PObject type" do
+        expected = parser.parse('Custom')
+        actual = parser.parse('Hash[String,Integer]')
+        expect(Puppet::Pops::Types::TypeMismatchDescriber::UnresolveTypeFinder).to receive(:new).and_return(unresolved_type_finder)
```
Comment (artlawson) [inline spec/unit/pops/types/type_mismatch_describer_spec.rb:236]: @joshcooper This line errors with 
uninitialized constant (Puppet::Pops::Types::TypeMismatchDescriber::UnresolveTypeFinder) 
Previously, the expect on line 237 wasn't recognizing the `Custom` type and we tried using 188 to fix that.

Comment (lucywyman) [inline spec/unit/pops/types/type_mismatch_describer_spec.rb:236]:
```suggestion
        expect(Puppet::Pops::Types::TypeMismatchDescriber::UnresolvedTypeFinder).to receive(:new).and_return(unresolved_type_finder)
```
💀 I knew it was a small detail.

Comment (joshcooper) [reply spec/unit/pops/types/type_mismatch_describer_spec.rb:236]: ah there's a typo, should be `Puppet::Pops::Types::TypeMismatchDescriber::UnresolvedTypeFinder`, note the `Unresolved` vs `Unresolve`

**Comment by lucywyman** [comment]
Hmm, I'm not sure how we got to 0 files changed....but regardless I think we're going to table this for now and potentially come back to it later https://github.com/puppetlabs/bolt/issues/2853#issuecomment-877283160.

**Pull request closed without merging**